test/unit/search/search_scheme_algorithm_test.cpp

```diff
@@ -0,0 +1,329 @@
+        EXPECT_LE(error_distribution[block], single_block_length);
+        if (error_distribution[block] > single_block_length)
+        {
+            debug_stream << "Error in block " << block << "(+ 1): " << error_distribution[block]
+                         << " errors cannot fit into a block of length " << single_block_length << "." << '\n'
+                         << "Error Distribution: " << error_distribution << '\n';
+            exit(1);
+        }
+
+        // Choose random positions in the query sequence for substitutions. Repeat until all error positions are unique.
+        std::vector<uint8_t> error_positions(error_distribution[block]);
+        do
+        {
+            error_positions.clear();
+            for (uint8_t error = 0; error < error_distribution[block]; ++error)
+                error_positions.push_back(std::rand() % single_block_length);
+            std::sort(error_positions.begin(), error_positions.end());
+        } while (std::adjacent_find(error_positions.begin(), error_positions.end()) != error_positions.end());
+
+        // Construct query sequence with chosen error positions.
+        for (uint8_t error = 0; error < error_positions.size(); ++error)
+        {
+            uint64_t const pos = error_positions[error] + current_blocks_length;
+            // Decrease alphabet size by one because we don't want to replace query[pos], with the same character.
+            uint8_t new_rank = std::rand() % (alphabet_size_v<char_t> - 1);
+            // If it is a match now, it can't be the highest rank of the alphabet. Thus we set it to the highest rank.
+            if (new_rank == to_rank(query[pos]))
+                new_rank = alphabet_size_v<char_t> - 1;
+            assign_rank(query[pos], new_rank);
+        }
+        current_blocks_length += single_block_length;
+    }
+
+    std::vector<uint64_t> hits_trivial, hits_ss;
+
+    auto delegate_trivial = [&hits_trivial](auto const & it)
+    {
+        auto const & hits_tmp = it.locate();
+        hits_trivial.insert(hits_trivial.end(), hits_tmp.begin(), hits_tmp.end());
+    };
+
+    auto delegate_ss = [&hits_ss](auto const & it)
+    {
+        auto const & hits_tmp = it.locate();
+        hits_ss.insert(hits_ss.end(), hits_tmp.begin(), hits_tmp.end());
+    };
+
+    uint8_t const total        = search.u.back();
+    uint8_t const substitution = std::rand() % (total + 1);
+
+    detail::search_param error_left{total, substitution, 0, 0};
+
+    // Find all hits using search schemes.
+    detail::search_ss<false>(it, query, start_pos, start_pos + 1, 0, 0, true, search, blocks_length, error_left,
+                             delegate_ss);
+    // Find all hits using trivial backtracking.
+    detail::search_trivial<false>(it, query, 0, error_left, delegate_trivial);
+
+    // Eliminate hits that we are not interested in (based on the search and chosen error distribution)
+    hits_ss.erase(std::remove_if(hits_ss.begin(), hits_ss.end(),
```

Comment:
why erase *and* remove_if ?

Reply:
Remove_if doesn't remove the elements, it just moves them to the back of the range and returns the iterator of the first element to be removed.